OutOfSchool/OutOfSchool.WebApi.Tests/Validators/CustomAgeAttributeTests.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using NUnit.Framework;
+using OutOfSchool.Common.Validators;
+
+namespace OutOfSchool.WebApi.Tests.Validators;
+
+[TestFixture]
+public class CustomAgeAttributeTests
+{
+    [Test]
+    public void IsValid_WhenDateIsNull_ShouldReturnTrue()
+    {
+        // Arrange
+        var dateOfBirth = null as DateTime?;
+        var customDayOfBirthAttribute = new CustomAgeAttribute();
+
+        // Act
+        var isValid = customDayOfBirthAttribute.IsValid(dateOfBirth);
+
+        // Assert
+        Assert.IsTrue(isValid);
+    }
+
+    [Test]
+    public void IsValid_WhenDateIsNotDateTimeOrDateOnly_ShouldReturnFalse()
+    {
+        // Arrange
+        var dateOfBirth = "1999-01-01";
+        var customDayOfBirthAttribute = new CustomAgeAttribute();
+
+        // Act
+        var isValid = customDayOfBirthAttribute.IsValid(dateOfBirth);
+
+        // Assert
+        Assert.IsFalse(isValid);
+    }
+
+    [Test]
+    public void IsValid_WhenDateIsDateTime_ShouldReturnTrue()
+    {
+        // Arrange
+        var dateOfBirth = DateTime.UtcNow;
+        var customDayOfBirthAttribute = new CustomAgeAttribute();
+
+        // Act
+        var isValid = customDayOfBirthAttribute.IsValid(dateOfBirth);
+
+        // Assert
+        Assert.IsTrue(isValid);
+    }
+
+    [Test]
+    public void IsValid_WhenDateIsDateOnly_ShouldReturnTrue()
+    {
+        // Arrange
+        var dateOfBirth = DateOnly.FromDateTime(DateTime.UtcNow);
+        var customDayOfBirthAttribute = new CustomAgeAttribute();
+
+        // Act
+        var isValid = customDayOfBirthAttribute.IsValid(dateOfBirth);
+
+        // Assert
+        Assert.IsTrue(isValid);
+    }
+
+    [Test]
+    public void IsValid_WhenMinAgeIsNotPositive_ShouldThrowInvalidOperationException()
+    {
+        // Arrange
+        var dateOfBirth = DateTime.UtcNow;
```

Comment:
(Test design) It doesn't matter what specific date is passed to the test. Please state it in your code. 
I.e. `var dateOfBirth = It.IsAny<DateTime>();`.

Also, as it is simple code and not reused, you could inline it in `.IsValid()` call

Reply:
This comment is also related to other test where specific date is not important.

Reply:
> (Test design) It doesn't matter what specific date is passed to the test. Please state it in your code. I.e. `var dateOfBirth = It.IsAny<DateTime>();`.
> 
> Also, as it is simple code and not reused, you could inline it in `.IsValid()` call

Personally, I prefer not use `It.IsAny<T>()` from `Moq` as variable assignment, I think this is considered to use only in setup the method invocation (eg `.Setup(..)`, `.Verify(..)` etc). But I can be wrong so can you explain why using `It.IsAny<DateTime>();` is preffered over using some value or even default keyword ?

It class docs: https://documentation.help/Moq/FBE0FFA5.htm

Reply:
> ... why using It.IsAny<DateTime>(); is preffered over using some value or even default keyword ?

"Some" value or "default" value are specific values. When I read test code I could think that test will work only for this specific value, but not for others (especially when you use one of the most specific values for `DateTime` - `UtcNow`).

For this test `DateTime` value doesn't matter. It could be any. `It.IsAny` is one of the way to literally specify this fact in test code. 

The other way - use random date, i.e. from fake data generator.  Test data generators - is better approach, but I see more projects without them, than projects having them, that's why `It.IsAny` - is in general a cheaper way. 

For this specific project - we have `Bogus` and you could use it as well. It's up to you.

Btw, there is other comment regarding `maxAge` check implementation and related test, and lack in test logic introduced by using the very specific input data.

Reply:
> > ... why using It.IsAny(); is preffered over using some value or even default keyword ?
> 
> "Some" value or "default" value are specific values. When I read test code I could think that test will work only for this specific value, but not for others (especially when you use one of the most specific values for `DateTime` - `UtcNow`).
> 
> For this test `DateTime` value doesn't matter. It could be any. `It.IsAny` is one of the way to literally specify this fact in test code.
> 
> The other way - use random date, i.e. from fake data generator. Using of test data generators - is better approach, but I've seen more projects without them, than projects having them, that's why `It.IsAny` - is in general a cheaper way.
> 
> For this specific project we have `Bogus` and you could use it as well. It's up to you.
> 
> Btw, there is other comment regarding `maxAge` check implementation and related test, and gap in test logic introduced by using the very specific input data.

Thanks, in my case I would consider using some fake generated date from `Bogus` rather than using `It.IsAny`.

Reply:
> ... is it okay ... have different seeds for bogus randomizer ?

The only profit - you test more variants in one test run. The downside - test suit run takes longer time (4 runs should be done for all tests, even for tests having no usage of Bogus); more test code has to be maintained (the seeding code); you would never get random data each time it will deterministic 4 specified, so the whole idea of randomness is lost.

If I were you I wouldn't use seeding at all. 
Bogus by default uses different seeds (and thus provides different random values) from one test run to another, so sooner or later problematic data would be found if it exists.

Reply:
> > ... is it okay ... have different seeds for bogus randomizer ?
> 
> The only profit - you test more variants in one test run. The downside - test suit run takes longer time (4 runs should be done for all tests, even for tests having no usage of Bogus); more test code has to be maintained (the seeding code); you would never get random data each time it will deterministic 4 specified, so the whole idea of randomness is lost.
> 
> If I were you I wouldn't use seeding at all. Bogus by default uses different seeds (and thus provides different random values) from one test run to another, so sooner or later problematic data would be found if it exists.

I remove seeding from tests, and run these tests until failure, and after finding some bugs and fixing them, everything is works fine (or maybe something would be founded later)